File: android/MyPlaygroundInJava/app/src/main/java/com/gmail/htaihm/myplaygroundinjava/MainActivity.java
package com.gmail.htaihm.myplaygroundinjava;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class MainActivity extends AppCompatActivity {

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);

    RecyclerView rvMainLlist = findViewById(R.id.rvMainList);
    rvMainLlist.setLayoutManager(new LinearLayoutManager(this));
    rvMainLlist.setAdapter(new MainListItemAdapter());
  }

  class MainListItemAdapter extends RecyclerView.Adapter<MainListItemViewHolder> {

    @NonNull
    @Override
    public MainListItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
      LayoutInflater layoutInflater = LayoutInflater.from(MainActivity.this);
      View view = layoutInflater.inflate(R.layout.main_list_item, parent, false);
      return new MainListItemViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull MainListItemViewHolder holder, int position) {
      MainListItem mainListItem = MainListItems.MAIN_LIST_ITEMS.get(position);
      holder.bindItem(mainListItem);
    }

    @Override
    public int getItemCount() {
      return MainListItems.MAIN_LIST_ITEMS.size();
    }
  }

  class MainListItemViewHolder extends RecyclerView.ViewHolder implements  View.OnClickListener{

    private TextView mTvMainListItem;
    private MainListItem mMainListItem;

    public MainListItemViewHolder(@NonNull View itemView) {
      super(itemView);

      mTvMainListItem = itemView.findViewById(R.id.tvMainListItem);
      itemView.setOnClickListener(this);
    }

    public void bindItem(MainListItem mainListItem) {
      mTvMainListItem.setText(mainListItem.getTitle());
      mMainListItem = mainListItem;
    }

    @Override
    public void onClick(View v) {
      startActivity(mMainListItem.createIntent(MainActivity.this));
    }
  }

}

File: android/jetpack/PlayDataBinding/app/src/main/java/com/gmail/htaihm/playdatabinding/MainActivity.kt
package com.gmail.htaihm.playdatabinding

import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.Log

/**
 * Reference: http://imakeanapp.com/android-jetpack-data-binding/
 *
 * Demo: data binding, binding adapters, observables
 */
class MainActivity : AppCompatActivity(), HeroesAdapter.OnHeroClickListener {

    private lateinit var recyclerView: RecyclerView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        recyclerView = findViewById(R.id.heroes_list)
        recyclerView.layoutManager = LinearLayoutManager(this)

        val repository = HeroesRepository.getInstance(applicationContext)
        val heroes = repository.fetchHeroes()

        val adapter = HeroesAdapter(heroes, this)
        recyclerView.adapter = adapter
    }

    override fun onHeroClick(hero: Hero) {
        Log.d("MainActivity", "Clicked Hero: ${hero.name}")
        startActivity(HeroActivity.createIntent(this, hero.id))
    }
}

